src/ui/__tests__/ui_app.js

```diff
@@ -0,0 +1,253 @@
+      const spec = getMockSpec({ testSpec: true });
+      let newApp;
+      beforeEach(() => {
+        newApp = new UiApp(uiExports, spec);
+      });
+
+      it('has a reference to the uiExports object', () => {
+        expect(newApp.uiExports).to.be(uiExports);
+      });
+
+      it('has a reference to the original spec', () => {
+        expect(newApp.spec).to.be(spec);
+      });
+
+      it('has a reference to the spec.injectVars function', () => {
+        const helloFunction = () => 'hello';
+        const spec = {
+          id: 'uiapp-test',
+          injectVars: helloFunction
+        };
+        const newApp = new UiApp(uiExports, spec);
+        expect(newApp.getInjectedVars).to.be(helloFunction);
+      });
+    });
+
+    describe('app.getInjectedVars', () => {
+      it('is noop function by default', () => {
+        const spec = {
+          id: 'uiapp-test'
+        };
+        const newApp = new UiApp(uiExports, spec);
+        expect(newApp.getInjectedVars).to.be(noop);
+      });
+    });
+
+    describe('hidden and listed flags', () => {
+      it('has hidden defaulted to false and listed defaulted to true', () => {
+        const spec = {
+          id: 'uiapp-test'
+        };
+        const newApp = new UiApp(uiExports, spec);
+        expect(newApp.hidden).to.be(false);
+        expect(newApp.listed).to.be(true);
+      });
+
+      it('has listed set to false if listed is passed as null and hidden is passed as true, ', () => {
+        const spec = {
+          id: 'uiapp-test',
+          hidden: true,
+          listed: null
+        };
+        const newApp = new UiApp(uiExports, spec);
+        expect(newApp.hidden).to.be(true);
+        expect(newApp.listed).to.be(false);
+      });
+
+      it('has listed set to true if listed is passed as null and hidden passed as false, ', () => {
+        const spec = {
+          id: 'uiapp-test',
+          hidden: false,
+          listed: null
+        };
+        const newApp = new UiApp(uiExports, spec);
+        expect(newApp.hidden).to.be(false);
+        expect(newApp.listed).to.be(true);
+      });
+
+      it('has listed set to true if listed is passed as null and hidden not passed', () => {
+        const spec = {
+          id: 'uiapp-test',
+          listed: null
+        };
+        const newApp = new UiApp(uiExports, spec);
+        expect(newApp.hidden).to.be(false);
+        expect(newApp.listed).to.be(true);
+      });
+
+      it('has listed set to what it was passed as', () => {
+        const spec = {
+          id: 'uiapp-test',
+          listed: true
+        };
+        const newApp = new UiApp(uiExports, spec);
+        expect(newApp.hidden).to.be(false);
+        expect(newApp.listed).to.be(true);
+
+        const spec2 = {
+          id: 'uiapp-test',
+          listed: false
+        };
+        const newApp2 = new UiApp(uiExports, spec2);
+        expect(newApp2.hidden).to.be(false);
+        expect(newApp2.listed).to.be(false);
```

Comment:
I disagree that it is tricky to figure out what went wrong if there is a failure because the failure will have a stack trace that points to the failing assertion. In the past I have went through some hoops to avoid that kind of failure messaging, but eventually just simplified because my team member and myself felt it wasn't worth the extra code.